# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')


In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns
df.head (5)

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'Exited']].copy()
df

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       France  Female   42       2       0.00              1          1   
1        Spain  Female   41       1   83807.86              1          0   
2       France  Female   42       8  159660.80              3          1   
3       France  Female   39       1       0.00              2          0   
4        Spain  Female   43       2  125510.82              1          1   
...        ...     ...  ...     ...        ...            ...        ...   
9995    France    Male   39       5       0.00              2          1   
9996    France    Male   35      10   57369.61              1          1   
9997    France  Female   36       7       0.00              1          0   
9998   Germany    Male   42       3   75075.31              2          1   
9999    France  Female   28       4  130142.79              1          1   

      IsActiveMember  Exited  
0                  1       1  
1                  1       0  
2                  0       1  
3                  0       0  
4                  1       0  
...              ...     ...  
9995               0       0  
9996               1       0  
9997               1       1  
9998               0       1  
9999               0       0  

[10000 rows x 9 columns]

## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')
dummies = pd.get_dummies(df['Geography'], ['Gender'],drop_first = True)
df = pd.concat([df,dummies],axis = 1).drop('Geography', axis = 1).drop('Gender', axis = 1)
df

RowNumber  CustomerId  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             1    15634602   42       2       0.00              1          1   
1             2    15647311   41       1   83807.86              1          0   
2             3    15619304   42       8  159660.80              3          1   
3             4    15701354   39       1       0.00              2          0   
4             5    15737888   43       2  125510.82              1          1   
...         ...         ...  ...     ...        ...            ...        ...   
9995       9996    15606229   39       5       0.00              2          1   
9996       9997    15569892   35      10   57369.61              1          1   
9997       9998    15584532   36       7       0.00              1          0   
9998       9999    15682355   42       3   75075.31              2          1   
9999      10000    15628319   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  ['Gender']_Germany  \
0                  1        101348.88       1                   0   
1                  1        112542.58       0                   0   
2                  0        113931.57       1                   0   
3                  0         93826.63       0                   0   
4                  1         79084.10       0                   0   
...              ...              ...     ...                 ...   
9995               0         96270.64       0                   0   
9996               1        101699.77       0                   0   
9997               1         42085.58       1                   0   
9998               0         92888.52       1                   1   
9999               0         38190.78       0                   0   

      ['Gender']_Spain  
0                    0  
1                    1  
2                    0  
3                    0  
4                    1  
...                ...  
9995                 0  
9996                 0  
9997                 0  
9998                 0  
9999                 0  

[10000 rows x 12 columns]

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
#df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')
X = df.drop ('Exited', axis = 1)
y = df ['Exited']


In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

### Jawab:
Model pertama yang digunakan adalah logistic regresion, dengan parameter tol dan c. tol merupakan singkatan dari tolerance, merupakan parameter untuk mengukur convergence criterion. C merupakan parameter regulasi yang digunakan untuk mencegah overfitting dan mengontrol kompleksitas model. Logistic regresion menggunakan fungsi logistik (sigmoid) untuk mengubah keluaran linear menjadi probabilitas kelas positif (1) dan kelas negatif (0).

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

from sklearn.linear_model import LogisticRegression
#START CODE
model1 = LogisticRegression()
params = {
    'tol': [0.0001,0.001,0.01,0,1],
    'C':[0.001,0.01,0.1,1,10,100]
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
best_params = grid.best_params_
print("Hyperparameter Terbaik:", best_params)

Hyperparameter Terbaik: {'C': 1, 'tol': 1}


In [ ]:
from sklearn.metrics._plot.confusion_matrix import ConfusionMatrixDisplay
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("Array")
print(confusion_matrix(y_test,y_pred))
print("Accuracy Score")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.98      0.89      1983
           1       0.67      0.18      0.29       517

    accuracy                           0.81      2500
   macro avg       0.75      0.58      0.59      2500
weighted avg       0.79      0.81      0.77      2500

Array
[[1937   46]
 [ 422   95]]
Accuracy Score
0.8128


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!
### Jawab:
Model kedua yang digunakan adalah Support Vector Machine, dengan parameter C, gamma, dan kernal. Parameter C untuk mengontrol margin maksimum dan jumlah kesalahan klasifikasi pada data pelatihan. Parameter kernel digunakan untuk menentukan jenis fungsi kernel yang akan digunakan. SVM merupakan algoritma yang digunakan untuk tugas klasifikasi dan regresi, yang mampu menangani data yang kompleks dan pemisahan kelas yang tidak linear.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

from sklearn.svm import SVC

#START CODE
model2 = SVC(kernel='linear',C=10)
params = {
    'C': [0.001, 0.01, 1, 10],
    'gamma': ['scale', 'auto'],
    'kernel': ['linear', 'rbf']
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model2,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
best_params = grid.best_params_
print("Hyperparameter terbaik Support Vector Machine:")
print(best_params)

Hyperparameter terbaik Support Vector Machine:
{'C': 10, 'gamma': 'scale', 'kernel': 'rbf'}


In [ ]:
from sklearn.metrics._plot.confusion_matrix import ConfusionMatrixDisplay
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("Array")
print(confusion_matrix(y_test,y_pred))
print("Accuracy Score")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.98      0.92      1983
           1       0.84      0.40      0.54       517

    accuracy                           0.86      2500
   macro avg       0.85      0.69      0.73      2500
weighted avg       0.86      0.86      0.84      2500

Array
[[1942   41]
 [ 309  208]]
Accuracy Score
0.86


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

### Jawab:
Pada model ketiga digunakan Random forest, dengan parameter n estimator, max depth, min sample split, dan min sample leaf. n estimator untuk mengontrol jumlah pohon keputusan yang akan dibangun. Max depth untukmembatasi tingkat kompleksitas setiap pohon, yang mana menunjukan kedalaman maksimum untuk setiap pohon keputusan. min sample split merupakan jumlah minimum sample yang diperlukan untuk membagi simpul dalam pohon keputusan. min sample leaf merupakan jumlah minimum yang diperlukan untuk menjadi daun (leaf) dalam pohon keputusan. Random forest merupakan algoritma ensemble dalam machine learning untuk klasifikasi dan regresi. Random forest menggabungkan prediksi dari beberapa model pohon keputusan (decision tree) untuk menghasilkan modek yang lebih stabil.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

from sklearn.ensemble import RandomForestClassifier
#START CODE
model3 = RandomForestClassifier()
params = {
    'n_estimators': [50, 100, 200],
    'max_depth': [5, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model3,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train, y_train)
best_params = grid.best_params_
print("Hyperparameter terbaik Random Forest:")
print(best_params)

grid.fit(X_train,y_train)
grid.best_params_

Hyperparameter terbaik Random Forest:
{'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 10, 'n_estimators': 100}


{'max_depth': 20,
 'min_samples_leaf': 4,
 'min_samples_split': 5,
 'n_estimators': 50}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("Array")
print(confusion_matrix(y_test,y_pred))
print("Accuracy")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.79      0.46      0.58       517

    accuracy                           0.86      2500
   macro avg       0.83      0.72      0.75      2500
weighted avg       0.86      0.86      0.85      2500

Array
[[1918   65]
 [ 277  240]]
Accuracy
0.8632


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Berdasarkan tiga model yang digunakan, didapatkan nilai akurasi Logistic Regresion 81,28%, SVM 86%, dan Random Forest 86,32%. Dilihat dari nilai akurasi yang didapatkan, model yang terbaikadalah Random Forest dengan akurasi 86,32%. Hal tersebut dikarenakan random forest memiliki tunnng yang lebih daripada SVM dan logistic regresion. Selai itu, random forest merupakan algoritma yang menggabungkan beberapa pohon keputusan yang mana mengurangi overfitting dan dapat meningkatkan hasil akurasi.